cmakelib/lexer/lexer_test.go

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lexer
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/alecthomas/participle/lexer"
+	plex "github.com/alecthomas/participle/lexer"
+	"github.com/alecthomas/repr"
+)
+
+type Token = plex.Token
+
+func NewToken(kind rune, value string) Token {
+	return Token{Type: kind, Value: value}
+}
+
+func NewTokenAt(kind rune, value string, offset, line, col int) Token {
+	return Token{Type: kind,
+		Value: value,
+		Pos: lexer.Position{
+			Filename: "",
+			Offset:   offset,
+			Line:     line,
+			Column:   col,
+		},
+	}
+}
+
+func compareTokens(a, b []Token) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for n, _ := range a {
+		if a[n].Type != b[n].Type || a[n].Value != b[n].Value {
+			return false
+		}
+	}
+	return true
+}
+
+func lexString(value string) ([]Token, error) {
+	lexer, err := New().Lex(strings.NewReader(value))
+	if err != nil {
+		return nil, err
+	}
+	return plex.ConsumeAll(lexer)
+}
+
+func removeWhitespace(toks []lexer.Token) []lexer.Token {
+	var r []lexer.Token
+	for _, tok := range toks {
+		switch tok.Type {
+		case Space, Newline:
+			continue
+		default:
+			r = append(r, tok)
+		}
+	}
+	return r
+}
+
+func TestDirectiveSpacing(t *testing.T) {
+	input := []string{
+		" directive ( ) ",
+		"\ndirective ( ) ",
+		"\ndirective\n( ) ",
+		"\ndirective\n(\n) ",
+		"\ndirective\n(\n)\n",
+		"directive\n(\n)\n",
+		"directive(\n)\n",
+		"directive()\n",
+		"directive()",
+		"directive #\n()",
+		"#\ndirective()",
+		"directive(#\n)",
+		"directive(#[[comment]])",
+		"#[[comment]]directive()",
+		"directive#[[comment]]()",
+		"directive#[[comment\n]]()",
+	}
+
+	expected := []Token{
+		{Type: Unquoted, Value: "directive"},
+		{Type: Punct, Value: "("},
+		{Type: Punct, Value: ")"},
+		plex.EOFToken(plex.Position{}),
+	}
+
+	for _, test := range input {
+		tokens, err := lexString(test)
+		if err != nil {
+			t.Errorf("Error lexing %#v: %s", test, err)
+			continue
+		}
+		tokens = removeWhitespace(tokens)
+		if len(tokens) != len(expected) {
+			t.Errorf("Invalid lex (%#v): %v != %v", test, expected, tokens)
```

Comment:
Conventionally Go puts the actual first, then the expected ("got, want", though those words aren't necessary).